Question 1 (#197)
UWB-ACM/CTCI: 2020-02-Spring/0_arrays_matrix/README.md

<a name="p1"/>

### 1. Best Time to Buy and Sell Stock

Source: [LeetCode](https://leetcode.com/problems/best-time-to-buy-and-sell-stock/)

#### Scenario

> Say you have an array of n elements representing stock prices.

Reviewer: Seems like and interesting problem. I'll submit a full review after you complete the question. Be sure to write down in the problem statement what format you want the output in, and whether there are any guarantees on the input set (i.e. no negatives, no nulls, w/e)

Author: I will finish it up first thing tomorrow. Thanks.

Price of the stock on the first day is the first element, second day is the second element and so on. 

> If you were only permitted to complete at most one transaction (i.e., buy one and sell one share of the stock), design an algorithm to find the maximum profit.

Reviewer 2: Here, "one transaction" is a bit confusing, since I tend to think of a transaction as **either** _buying_ **or** _selling_. Leaving a suggestion to say "two transactions" -- feel free to come up with something better which explains you must buy and then sell the stock to achieve profit (or minimize loss).

```suggestion
If you were only permitted to complete at most two transactions (i.e., buy one and sell one share of the stock), design an algorithm to find the maximum profit.
```

Reviewer: Seconded on 2 transactions

Note that you cannot sell a stock before you buy one.

Note that you can only sell forwards, not backwards. 

#### Example Input

**Input:** \[7,1,5,3,6,4]

> **Output:** 5

Reviewer: So outputting the purchase date doesn't matter? Why? If it doesn't, make that clear in the problem statement.

Author: I had to modify the question a bit because the original was kinda confusing. Should I make it more obvious or should I just leave it vague like that so people can kinda practice asking clarifying questions?

Reviewer 2: I personally think this is fine as-is -- currently, the problem statement says "find the maximum profit", but the reader must clarify (or infer) that it is the maximum price difference for a single unit of stock. This isn't inconsistent and is an acceptable amount of vague imo.

For any array-type problem, an interviewer can create problem variations on the fly by changing the return value requirements. (ie "What if you had to return the dates instead of the maximum profit?"). It's up to y'all how you want to incorporate that idea into your sessions.

Reviewer: So the reason this is confusing, is because both the sell date and the max profit is 5. So I thought we were returning the date. I think it would be nice if we could change the example a bit, but it would also be OK to say this is a trick you're meant to catch/question.

**Explanation:** 
Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
Not 7-1 = 6, as selling price needs to be larger than buying price.

#### Function Signature

**Java:**

> ```

Reviewer 2: Syntax highlighting is nice

````suggestion
```java
````

class Solution {
    public int maxProfit(int[] prices) {
        
    }
}

```

**C++:**

> ```

Reviewer 2: Syntax highlighting is nice
````suggestion
```c++
````

<!-- Don't remove -->
<a name="s1"/>

### 1. Best Time to Buy and Sell Stock

Source: [LeetCode](https://leetcode.com/problems/best-time-to-buy-and-sell-stock/solution/)

#### Naive/Simple Solution

We need to find out the maximum difference (which will be the maximum profit) between two numbers in the given array. Also, the second number (selling price) must be after the first one (buying price) because you sell in the future, not in the past.

In other terms, we need to find max(prices[j]−prices[i]), for every i and j such that j > i.

**Brute Force:**
```

public int maxProfit(int prices[]) {
        int maxprofit = 0;
        for (int i = 0; i < prices.length - 1; i++) {
            for (int j = i + 1; j < prices.length; j++) {
                int profit = prices[j] - prices[i];
                if (profit > maxprofit)
                    maxprofit = profit;
            }
        }
        return maxprofit;
    }
    
Complexity Analysis

   Time complexity: O(n^2)
  
   Space complexity : O(1). Only two variables - maxprofit and profit are used.

```

#### Optimal Solution

> See [solution](/buy_stock/solution.java)

Reviewer 2: ```suggestion
See [solution](./buy_stock/solution.java)
```

#### Testing The Solutions OR Driver For Solution

> See [driver](/buy_stock/driver.java)

Reviewer 2: ```suggestion
See [driver](./buy_stock/driver.java)
```
